사용자 연속 요청 block구현 및 세션 청소기 구현

kimykstar opened the pull request:
## 작업 개요

- Resolves #247 
- Resolves #252 

## 작업 상세 내용

### RequestGuard구현
- 해당 Guard는 SessionTable에 저장된 사용자 별 lastRequest시간 정보를 기반으로 현재 요청이 `0.5초` 보다 더 짧은 간격의 요청이라면 막는 기능을 합니다.
- 해당 Guard는 사용자와 많은 상호작용이 있는 API에 대해서만 적용시켰습니다.(`command`)

### @nestjs/schedule을 활용한 세션 청소기 구현
- `Cron`을 활용하여 만료시간에 도달한 세션을 `10분`마다 주기적으로 삭제할 수 있도록 구현하였습니다.
![image](https://github.com/user-attachments/assets/154a7e84-6ed5-4107-87c9-b4d35a971ac3)
위의 이미지에 대한 설명입니다.
- 1번째 줄: 실행됐음을 알리는 그냥 제가 console.log('cron')으로 출력한 것입니다.
- 2번째 줄: 삭제 전 세션 테이블 출력입니다.
- 3번째 줄: 삭제 후 세션 테이블 출력입니다.
위의 상황은 `세션 만료 시간`은 1초이며, `Cron 주기`는 10초입니다.

## 참고자료(선택)

[nextjs/schedule](https://docs.nestjs.com/techniques/task-scheduling)
[개발 일지](https://github.com/boostcampwm-2024/web34-LearnDocker/wiki/%5B5%EC%A3%BC-3%EC%9D%BC%EC%B0%A8-%E2%80%90-J048-%EA%B9%80%EC%98%81%EA%B4%80%5D-%EA%B0%9C%EB%B0%9C-%EC%9D%BC%EC%A7%80%28%EC%97%B0%EC%86%8D%EC%9A%94%EC%B2%AD-%EC%B2%98%EB%A6%AC-%EB%B0%8F-%EC%84%B8%EC%85%98%EA%B4%80%EB%A6%AC%29)

## 문제점 혹은 고민(선택)
- 연속 요청 시 `lastRequest`갱신의 시점에 대한 고민입니다.
  1. block되지 않은 시점에만 lastRequest를 갱신한다.
  2. block된 시점에도 lastRequest를 갱신한다.(이 방식은 block되어도 갱신되기 때문에 요청 시간간격이 계속 짧다면 계속 block됩니다.)
- 사용자가 탭을 닫을 때 이벤트를 처리해서 구현을 해보았는데 이벤트 핸들러가 잘 동작하지 않습니다... ㅠ `beforeunload`이벤트에 대해서 등록을 하고 `e.preventDefault()`함수를 적용시켜 닫기지 않도록 막아도 그냥 닫혀버립니다..

kimykstar committed:
feat: RequestGuard구현

kimykstar committed:
feat: RequestGuard 컨트롤러에 등록

- 학습 시작버튼을 누를 시 start, endDate요청이 연속 2번이 발생함으로
  RequestGuard에 의해 요청이 처리되지 않음
- RequestGuard에 걸리지 않도록 클라이언트에 start요청에 대한 응답으로
  endDate를 준다.

kimykstar committed:
feat: api요청 연속 발생 제거

kimykstar committed:
feat: statuscode적용 및 Guard적용api변경

- 사용자와 많은 상호작용 이후 전달되는 API에 대해서만 Guard적용
- TooManyRequestsException추가 및 처리
- RequestGuard동작 변경

kimykstar committed:
feat: Alert창을 활용해 429상태코드 처리

- showAlert함수 props로 전달
- 429상태코드에 대해서 Alert창으로 표시

kimykstar committed:
config: @nestjs/schedule패키지 설치

- 세션 청소기를 돌리기 위한 패키지 설치

kimykstar committed:
feat: 10분 주기마다 세션 정보 삭제 청소기 구현

- @nestjs/schedule패키지를 활용해 10분마다 주기적으로 세션 정보 삭제

tuchongkim commented on backend/src/quiz/quiz.module.ts:
```diff
@@ -5,9 +5,10 @@ import { TypeOrmModule } from '@nestjs/typeorm';
+import { RequestModule } from 'src/common/request/request.module';
 
 @Module({
-    imports: [TypeOrmModule.forFeature([Quiz]), SandboxModule, CacheModule],
+    imports: [TypeOrmModule.forFeature([Quiz]), SandboxModule, CacheModule, RequestModule],
     controllers: [QuizController],
```
혹시 모듈 부분 궁금하셨던 부분이 여기일까요? ReuestModule에도 `@Global` 을 붙이면 import 하지 않아도 됩니다!

kimykstar replied:
감삼다!

tuchongkim commented on backend/src/common/request/request.service.ts:
```diff
@@ -0,0 +1,25 @@
+    validRequestInterval(sessionId: string) {
+        const sessionDatas = this.cacheService.get(sessionId) as UserSession;
+        const prevReqTime = sessionDatas.lastRequest.getTime();
+        const currentReqTime = new Date().getTime();
+        const interval = currentReqTime - prevReqTime;
+        if (interval < LIMIT) {
+            throw new RequestIntervalException();
+        }
+        this.cacheService.set(sessionId, {
+            ...sessionDatas,
+            lastRequest: new Date(),
+        });
```
new Date()를 2번 사용하신 이유가 있을까요?

kimykstar replied:
아 currentReqTime에서 new Date와 lastRequest에서의 new Date를 말씀하시는 것 같습니다! 위 new Date()의 경우에는 .getTime()을 통해 `number`타입의 시간 데이터를 저장해야 하고, 아래 new Date에서는 현재 시간의 `Date`타입의 시간 데이터를 저장하려고 했습니다. 기능 동작에만 신경쓰다보니 객체를 두번 만들 필요는 없을 것 같네요! ㅎㅎ 하나의 객체로 데이터를 저장하도록 수정하겠습니다.

SH9480P commented on backend/src/app.module.ts:
```diff
@@ -37,6 +39,8 @@ import { LoggerMiddleware } from './common/logger/middleware';
             renderPath: '/*',
         }),
         AuthModule,
+        CacheModule,
```
CacheModule은 제거해도 잘 동작하네요~
CacheModule의 service가 RequestModule에서 사용되기 위해서 RequestModule에서 CacheModule을 import 하셨죠?
AppModule에서는 직접 사용되고 있지 않기 때문에 여기서는 CacheModule을 import할 필요는 없습니다!
(라고 알고 있습니다)

kimykstar replied:
감사합니다!

SH9480P commented on backend/src/common/cache/cache.service.ts:
```diff
@@ -24,4 +26,13 @@ export class CacheService {
         this.store.delete(key);
     }
+
+    @Cron(CronExpression.EVERY_10_MINUTES)
+    deleteExpiredSession() {
+        this.store.forEach((values, sessionId) => {
+            if (new Date(values.startTime).getTime() + SESSION_DURATION < new Date().getTime()) {
+                this.delete(sessionId);
+            }
+        });
+    }
```
호스트 컨테이너도 지워주어야 합니다!

guune commented on frontend/src/components/StartButton.tsx:
```diff
@@ -16,9 +13,8 @@ const StartButton = (props: StartButtonProps) => {
-        const success = await createHostContainer(navigate);
-        const { endDate } = (await requestExpirationTime()) as ExpirationTime;
-        if (success) {
+        const endDate = await createHostContainer(navigate);
```
endDate 변수가 궁금합니다. 무슨 의미인가요?

kimykstar replied:
세션 만료 시간입니다! 처음에 maxAge로 했다가 변경했습니다!

tuchongkim commented on backend/src/sandbox/sandbox.controller.ts:
```diff
@@ -45,8 +46,10 @@ export class SandboxController {
-            const newSessionId = await this.sandboxService.assignContainer(ipAddress);
-            res.cookie('sid', newSessionId, { httpOnly: true, maxAge: SESSION_DURATION });
+            const sessionData = await this.sandboxService.assignContainer(ipAddress);
+            const startTime = sessionData?.startTime as Date;
+            res.cookie('sid', sessionData?.sessionId, { httpOnly: true, maxAge: SESSION_DURATION });
+            res.json({ endDate: new Date(startTime).getTime() + SESSION_DURATION });
```
혹시 sandbox/command 엔드포인트에서 원래 반환하던 데이터는 어떻게 처리되나요?

tuchongkim replied:
다음부터 더 꼼꼼히 확인하겠습니다..하하..

kimykstar replied:
하하 저도 몰랐던 사실입니다 ㅎㅎㅎ

kimykstar committed:
feat: 세션 청소기 호스트 컨테이너 삭제 로직 추가

- 호스트 컨테이너를 삭제하는 로직을 청소기에 추가함
- RequestModule을 Global로 설정하여 모듈 의존성 설정
- Date객체 1개만 생성하여 활용하도록 변경
